stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,240 @@
+#ifndef STAN_MATH_REV_FUNCTOR_REDUCE_SUM_HPP
+#define STAN_MATH_REV_FUNCTOR_REDUCE_SUM_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/functor.hpp>
+#include <stan/math/rev/core.hpp>
+#include <tbb/task_arena.h>
+#include <tbb/parallel_reduce.h>
+#include <tbb/blocked_range.h>
+
+#include <tuple>
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Var specialization of reduce_sum_impl
+ *
+ * @tparam ReduceFunction Type of reducer function
+ * @tparam ReturnType Must be var
+ * @tparam Vec Type of sliced argument
+ * @tparam Args Types of shared arguments
+ */
+template <typename ReduceFunction, typename ReturnType, typename Vec,
+          typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
+                       Vec, Args...> {
```

Comment:
I am afraid that this specialisation magic is not working right now correctly. I just ran the ode test model with this branch and when using reduce_sum I am not getting any speedup and Stan is also not using multiple threads.

... and a test for this would be great ... not yet sure how to test that though.

Reply:
I'm a dumb dumb @bbbales2 I forgot to add reduce_sum.hpp in rev to rev/functor.hpp

Reply:
No worries, this can happen - let's just put a test as suggested below to make sure we will never deploy this type of mistake.